bindings/python/pymongocrypt/binding.py

```diff
@@ -0,0 +1,749 @@
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
+ */
+bool
+mongocrypt_kms_ctx_message (mongocrypt_kms_ctx_t *kms,
+                            mongocrypt_binary_t *msg);
+
+/**
+ * Get the hostname from which to connect over TLS.
+ *
+ * The storage for @p endpoint is not owned by the caller, but
+ * is valid until calling @ref mongocrypt_ctx_kms_done.
+ *
+ * @param[in] kms A @ref mongocrypt_kms_ctx_t.
+ * @param[out] endpoint The output hostname as a NULL terminated string.
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
+ */
+bool
+mongocrypt_kms_ctx_endpoint (mongocrypt_kms_ctx_t *kms, const char **endpoint);
+
+/**
+ * Indicates how many bytes to feed into @ref mongocrypt_kms_ctx_feed.
+ *
+ * @param[in] kms The @ref mongocrypt_kms_ctx_t.
+ * @returns The number of requested bytes.
+ */
+uint32_t
+mongocrypt_kms_ctx_bytes_needed (mongocrypt_kms_ctx_t *kms);
+
+/**
+ * Feed bytes from the HTTP response.
+ *
+ * Feeding more bytes than what has been returned in @ref
+ * mongocrypt_kms_ctx_bytes_needed is an error.
+ *
+ * @param[in] kms The @ref mongocrypt_kms_ctx_t.
+ * @param[in] bytes The bytes to feed.
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
+ */
+bool
+mongocrypt_kms_ctx_feed (mongocrypt_kms_ctx_t *kms, mongocrypt_binary_t *bytes);
+
+/**
+ * Get the status associated with a @ref mongocrypt_kms_ctx_t object.
+ *
+ * @param[in] kms The @ref mongocrypt_kms_ctx_t object.
+ * @param[out] status Receives the status.
+ *
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
+ */
+bool
+mongocrypt_kms_ctx_status (mongocrypt_kms_ctx_t *kms,
+                           mongocrypt_status_t *status);
+
+/**
+ * Call when done handling all KMS contexts.
+ *
+ * @param[in] ctx The @ref mongocrypt_ctx_t object.
+ *
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
+ */
+bool
+mongocrypt_ctx_kms_done (mongocrypt_ctx_t *ctx);
+
+/**
+ * Perform the final encryption or decryption.
+ *
+ * @param[in] ctx A @ref mongocrypt_ctx_t.
+ * @param[out] out The final BSON to send to the server.
+ *
+ * The lifetime of @p out is tied to the lifetime of @p ctx. It is valid
+ * until @ref mongocrypt_ctx_destroy is called.
+ *
+ * @returns a bool indicating success.
+ */
+bool
+mongocrypt_ctx_finalize (mongocrypt_ctx_t *ctx, mongocrypt_binary_t *out);
+
+/**
+ * Destroy and free all memory associated with a @ref mongocrypt_ctx_t.
+ *
+ * @param[in] ctx A @ref mongocrypt_ctx_t.
+ */
+void
+mongocrypt_ctx_destroy (mongocrypt_ctx_t *ctx);
+""")
+
+lib = ffi.dlopen("mongocrypt")
+
+
+def libmongocrypt_version():
+    """Returns the version of libmongocrypt."""
```

Comment:
Can you update the docstring to indicate that this is a convenience method added to avoid an additional call to `decode`? Also, can you say why a similar utility isn't added for `mongocrpyt_status_message`? At first glance, it appears it might be worthwhile to define a private utility to `decode` return values from functions bearing return type `* char *`.

```
def _sanitize_output(retval):
    assert isinstance(retval, bytes_type)     # bytes_type is str on PY2 defined in compat
    if PY3:
        return retval.decode()
    return retval
```

Reply:
libmongocrypt_version is a convenient method to get the version of libmongocrypt that the pymongocrypt library is using. It will be useful for people reporting bugs as described in the "How To Ask For Help" section of the readme. 

We may add a `_sanitize_output` like method in a future PR but it's not needed yet.

Reply:
Actually I just added a `_to_string()` method. It's useful because [`ffi.string`](https://cffi.readthedocs.io/en/release-1.12/ref.html#ffi-string-ffi-unpack) returns `str` on Python 2 and `bytes` on Python 3. `_to_string()` returns `str` on both Python 2 and 3.

Reply:
Nice.